crates/cext/src/circuit.rs

```diff
@@ -1008,14 +1011,29 @@ pub unsafe extern "C" fn qk_circuit_instruction_clear(inst: *mut CInstruction) {
 /// Behavior is undefined if ``op_counts`` is not the object returned by ``qk_circuit_count_ops``.
 #[no_mangle]
 #[cfg(feature = "cbinding")]
-pub unsafe extern "C" fn qk_opcounts_free(op_counts: OpCounts) {
-    // SAFETY: Loading data contained in OpCounts as a slice which was constructed from a Vec
-    let data = unsafe { std::slice::from_raw_parts_mut(op_counts.data, op_counts.len) };
-    let data = data.as_mut_ptr();
-    // SAFETY: Loading a box from the slice pointer created above
-    unsafe {
-        let _ = Box::from_raw(data);
+pub unsafe extern "C" fn qk_opcounts_clear(op_counts: *mut OpCounts) {
+    // SAFETY: The user guarantees the input is a valid OpCounts pointer.
+    let op_counts = unsafe { mut_ptr_as_ref(op_counts) };
+
+    if op_counts.len > 0 && !op_counts.data.is_null() {
+        // SAFETY: We load the box from a slice pointer created from
+        // the raw parts from the OpCounts::data attribute.
+        unsafe {
+            let vec = Box::from_raw(std::slice::from_raw_parts_mut(
+                op_counts.data,
+                op_counts.len,
+            ));
+            // free the allocated strings in each OpCount
+            for count in vec.iter() {
+                if !count.name.is_null() {
+                    let _ = CString::from_raw(count.name as *mut c_char);
+                }
+            }
+            // the variable vec goes out of bounds and is freed too
+        }
     }
+    op_counts.len = 0;
```

Comment:
I wonder if we really need this... I would think it's reasonable to assume that after calling `qk_opcounts_clear` the `QkOpCounts` struct is no longer meant to be used, so why should we bother resetting its fields?

Reply:
What concerned me here a bit is that the `QkOpCounts` is still valid to use after calling `qk_opcounts_clear` but in an inconsistent state -- the length is still the old value but the memory is cleared and not readable. I rather didn't want to make any assumptions about whether the user means to discard the object or not 😄

Reply:
I agree for `*_free`, this object is not safe to use anymore after -- but I'd say `free` implies that whereas `clear` to me means that it's just sort of reset, not unsafe to use. A bit related to Jake's posted message on https://github.com/git/git/blob/c44beea485f0f2feaf460e2ac87fdd5608d63cf0/Documentation/CodingGuidelines#L621-L637 😄